src/common/runtime/src/ratelimit.rs

```diff
@@ -0,0 +1,608 @@
+//!     .unwrap();
+//!
+//! // Another use case might be admission control, where we start with some
+//! // initial budget and replenish it periodically. In this example, our
+//! // ratelimiter allows 1000 tokens/hour. For every hour long sliding window,
+//! // no more than 1000 tokens can be acquired. But all tokens can be used in
+//! // a single burst. Additional calls to `try_wait()` will return an error
+//! // until the next token addition.
+//! //
+//! // This is popular approach with public API ratelimits.
+//! let ratelimiter = Ratelimiter::builder(1000, Duration::from_secs(3600))
+//!     .max_tokens(1000)
+//!     .initial_available(1000)
+//!     .build()
+//!     .unwrap();
+//!
+//! // For very high rates, we should avoid using too short of an interval due
+//! // to limits of system clock resolution. Instead, it's better to allow some
+//! // burst and add multiple tokens per interval. The resulting ratelimiter
+//! // here generates 50 million tokens/s and allows no more than 50 tokens to
+//! // be acquired in any 1 microsecond long window.
+//! let ratelimiter = Ratelimiter::builder(50, Duration::from_micros(1))
+//!     .max_tokens(50)
+//!     .build()
+//!     .unwrap();
+//!
+//! // constructs a ratelimiter that generates 100 tokens/s with no burst
+//! let ratelimiter = Ratelimiter::builder(1, Duration::from_millis(10))
+//!     .build()
+//!     .unwrap();
+//!
+//! for _ in 0..10 {
+//!     // a simple sleep-wait
+//!     if let Err(sleep) = ratelimiter.try_wait() {
+//!            std::thread::sleep(sleep);
+//!            continue;
+//!     }
+//!     
+//!     // do some ratelimited action here    
+//! }
+//! ```
+
+use core::sync::atomic::{AtomicU64, Ordering};
+use std::fmt::{Debug, Formatter};
+
+use clocksource::precise::{AtomicInstant, Duration, Instant};
+use parking_lot::RwLock;
+use snafu::Snafu;
+
+// #[derive(Error, Debug, PartialEq, Eq)]
+// pub enum Error {
+//     #[error("available tokens cannot be set higher than max tokens")]
+//     AvailableTokensTooHigh,
+//     #[error("max tokens cannot be less than the refill amount")]
+//     MaxTokensTooLow,
+//     #[error("refill amount cannot exceed the max tokens")]
+//     RefillAmountTooHigh,
+//     #[error("refill interval in nanoseconds exceeds maximum u64")]
+//     RefillIntervalTooLong,
+// }
+#[derive(Debug, Snafu, PartialEq, Eq)]
+pub enum Error {
+    #[snafu(display("available tokens cannot be set higher than max tokens"))]
+    AvailableTokensTooHigh,
+
+    #[snafu(display("max tokens cannot be less than the refill amount"))]
+    MaxTokensTooLow,
+
+    #[snafu(display("refill amount cannot exceed the max tokens"))]
+    RefillAmountTooHigh,
+
+    #[snafu(display("refill interval in nanoseconds exceeds maximum u64"))]
+    RefillIntervalTooLong,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+struct Parameters {
+    capacity: u64,
+    refill_amount: u64,
+    refill_interval: Duration,
+}
+
+pub struct Ratelimiter {
```

Comment:
We may just use the ratelimit crate instead of writing our own:https://docs.rs/ratelimit/0.7.1/ratelimit/